src/dos.rs

```diff
@@ -0,0 +1,68 @@
+use handlebars::Handlebars;
+use rust_embed::RustEmbed;
+use serde::Serialize;
+use std::fs;
+use std::io::Write;
+use std::path::Path;
+use yore::code_pages::CP437;
+
+#[derive(RustEmbed)]
+#[folder = "$CARGO_MANIFEST_DIR/templates/dos"]
+struct Asset;
+
+pub struct Templates<'a> {
+  hbars: Handlebars<'a>,
+}
+
+impl Templates<'_> {
+  pub fn new() -> Templates<'static> {
+      let mut hbars = Handlebars::new();
+      hbars.register_escape_fn(handlebars::no_escape);
+      return Templates { hbars: hbars };
+  }
+
+  pub fn render_string<T: Serialize>(&self, template: &String, vars: T) -> Result<String, String> {
+      return match self.hbars.render_template(template.as_str(), &vars) {
+          Ok(rendered) => Ok(rendered),
+          Err(e) => return Err(format!("Couldn't render template: {0}", e)),
+      };
+  }
+
+  pub fn render_template<T: Serialize>(&self, name: &str, vars: T) -> Result<String, String> {
+      let template = match Asset::get(format!("{0}.hbr", name).as_str()) {
+          Some(asset) => match String::from_utf8(asset.data.to_vec()) {
+              Ok(data) => data,
+              Err(e) => return Err(format!("Couldn't load template for {0}: {1}", name, e)),
+          },
+          None => return Err(format!("Couldn't find template for {0}", name)),
+      };
+      return self.render_string(&template, vars);
+  }
+
+  pub fn write_dos<T: Serialize>(&self, name: &str, dir: &Path, vars: T) -> Result<(), String> {
```

Comment:
Seconding the recommendation to use a proper error handling trait instead of composing and stacking strings. You'll get features like stacktraces and causative relationships. And if you write rich error structs with a `Display` implementation you can get that *and* still have pretty user-facing messages.